src/zen/websocket_server.cpp

```diff
@@ -0,0 +1,1360 @@
+        // will be referenced when shutting down in order not destroying ptr while executing this thread
+        boost::shared_ptr<WsHandler> thisRef;
+
+        this->t_id = t_id;
+
+        try
+        {
+            localWs.reset(new websocket::stream<tcp::socket> { std::move(socket) });
+            LogPrint("ws", "%s():%d - allocated localWs %p\n", __func__, __LINE__, localWs.get());
+
+            localWs->set_option(
+                websocket::stream_base::decorator(
+                    [](websocket::response_type& res)
+                        {
+                            res.set(http::field::server,
+                            std::string(BOOST_BEAST_VERSION_STRING) + " Horizen-sidechain-connector");
+                        }));
+
+            localWs->control_callback(
+                [](websocket::frame_type kind, boost::string_view payload)
+                {
+                    if (kind == websocket::frame_type::ping)
+                    {
+                        std::string payl(payload);
+                        LogPrint("ws", "%s():%d - ping received... payload[%s]\n", __func__, __LINE__, payl);
+                    }
+                    // Do something with the payload
+                    boost::ignore_unused(kind, payload);
+                });
+
+            localWs->accept();
+
+            std::thread write_t(&WsHandler::writeLoop, this);
+            readLoop();
+            exit_rwhandler_thread_flag = true;
+            write_t.join();
+            socket.close();
+        }
+        catch (boost::system::system_error const& se)
+        {
+            if (se.code() != websocket::error::closed)
+            {
+                LogPrint("ws", "%s():%d - boost error %s\n", __func__, __LINE__, se.code().message());
+            }
+            LogPrint("ws", "%s():%d - websocket close\n", __func__, __LINE__);
+        }
+        catch (std::exception const& e)
+        {
+            LogPrint("ws", "%s():%d - error: %s\n", __func__, __LINE__, std::string(e.what()));
+        }
+        LogPrint("ws", "%s():%d - exit thread final\n", __func__, __LINE__);
+        {
+            std::unique_lock<std::mutex> lck(wsmtx);
+            this->shutdown();
+            tot_connections--;
+            LogPrint("ws", "%s():%d - connection[%u] closed: tot[%d]\n", __func__, __LINE__, t_id, tot_connections);
+
+            auto it = listWsHandler.begin();
+            while (it != listWsHandler.end())
+            {
+                if (this == (*it).get() )
+                {
+                    thisRef = (*it);
+                    LogPrint("ws", "%s():%d - removing handler obj from list\n", __func__, __LINE__);
+                    listWsHandler.erase(it++);
+                }
+                else
+                {
+                    ++it;
+                }
+            }
+        }
+    }
+
+    void send_tip_update(int height, const std::string& strHash, const std::string& blockHex, const CBlockIndex * pindex)
```

Comment:
Consider using a const std::string& chainWorkStr instead of passing the pindex.
As commented above for sendBlockEvent(), the reason is that we already are passing other parameters such as: pindex->nHeight, pindex->GetBlockHash()
therefore we might as well pass pindex->nChainWork.GetHex().